src/main/groovy/org/pih/warehouse/core/identification/BlankIdentifierResolver.groovy

```diff
@@ -0,0 +1,76 @@
+package org.pih.warehouse.core.identification
+
+import groovy.util.logging.Slf4j
+import org.grails.datastore.gorm.GormEntity
+
+/**
+ * Handles generating identifiers for all instances of an entity that don't yet have one.
+ *
+ * Used by the AssignIdentifierJob as a backup retry mechanism to ensure all entities have a valid id.
+ *
+ * Classes that implement this one will almost certainly also extend from IdentifierService. The only reason this trait
+ * is kept separate from the base service is to allow services to opt out of providing the functionality if they're
+ * not called by the AssignIdentifierJob.
+ */
+@Slf4j
+trait BlankIdentifierResolver<T extends GormEntity> {
+
+    /**
+     * @return the domain-specific key that is used in identifier properties.
+     * Ex: The "product" of "openboxes.identifier.product.format"
+     */
+    abstract String getEntityKey()
+
+    /**
+     * @return A list of all entities that do not have an identifier assigned.
+     */
+    abstract List<T> getAllUnassignedEntities()
+
+    /**
+     * Sets the identifier field on the given entity instance.
+     */
+    abstract void setIdentifierOnEntity(String id, T entity)
+
+    /**
+     * Generates a new identifier for the entity using the given params.
+     *
+     * This implementation will almost certainly come from the IdentifierService.
+     */
+    abstract String generate(IdentifierGeneratorParams params=null)
+
+    /**
+     * Generates an identifier for the given entity.
+     * Can be overridden by child classes if they require additional logic when setting unassigned identifiers.
+     */
+    String generateForEntity(T entity) {
+        return generate(IdentifierGeneratorParams.builder()
+                .templateEntity(entity)
```

Comment:
I actually originally had the generate method passing in the entity as a requirement but it got complicated because there were some cases where two different entities were sharing the same format for generating ids so it got wonky with all the typing.

I was struggling with the balance of "this should be driven 100% via properties" and "well we need to provide certain params in certain situations directly in the code". With the custom params it's kinda inevitable that we need to provide them via code, but for everything else it can be determined purely via properties.

I suppose I was trying too hard to make the identifiers not need to rely on a specific entity (by making it optional), but given that the property config is entity specific, maybe we should just lean in and say an identifier service implementation is for a specific entity, full stop. Any entity that uses an identifier must define their own service.